lib/server.js

```diff
@@ -232,10 +232,20 @@ class Server extends KarmaEventEmitter {
         let newBrowser = info.id ? (capturedBrowsers.getById(info.id) || singleRunBrowsers.getById(info.id)) : null
 
         if (newBrowser) {
+          // By default if a browser disconnects while still executing, we assume that the test
+          // execution still continues because just the socket connection has been terminated. Now
+          // since we know whether this is just a socket reconnect or full client reconnect, we
+          // need to update the browser state accordingly. This is necessary because in case a
+          // browser crashed and has been restarted, we need to start with a fresh execution.
+          if (!info.isSocketReconnect) {
+            newBrowser.setState(Browser.STATE_DISCONNECTED)
+          }
+
           newBrowser.reconnect(socket)
 
-          // We are restarting a previously disconnected browser.
-          if (newBrowser.state === Browser.STATE_DISCONNECTED && config.singleRun) {
+          // In case we are in single run and the reconnected browser was not able to restore
```

Comment:
Rather than describing the conditional guard, can the comment explain *why* we guard execute() in this way?

Reply:
Done.